src/messages/MMgrDigest.h

```diff
@@ -49,6 +49,14 @@ class MMgrDigest : public Message {
 private:
   ~MMgrDigest() override {}
 
+  using RefCountedObject::put;
+  using RefCountedObject::get;
+  friend void intrusive_ptr_add_ref(const MMgrDigest* m) {
+    m->get();
+  }
+  friend void intrusive_ptr_release(const MMgrDigest* m) {
+    m->put();
+  }
```

Comment:
@batrick i admit this is not as elegant as `RefCountedObjectInstanceSafe`. but i think the upside of it is its simplicity and readability. and ... arguably its ugliness. i think, in the long run, we will stop repeating this boilerplate, and move it into the `Message` class.

but if you hold a strong opinion to this, probably we can have a `SafeMessage` which inherits from `Message`. and let the opt-in Message types to inherit from `SafeMessage`. `SafeMessage` will be an instance of `RefCountedObjectSubTypeSafe`. as you suggest in https://github.com/ceph/ceph/pull/27325/commits/9a3d118a49a7140e41b1c812490cf7ef352abfa6 . but i really want to avoid this degree of complicity if possible.

Reply:
I don't see how this achieves the objective 9a3d118, i.e. prevent the user of `MMgrDigest` from calling `->get` or `->put`.

Reply:
Oh, I see, you're making these aliases **private**.

Honestly, I think it is not that much complexity to have a simple helper class in `RefCountedObject` which deletes `get`/`put`. Alternatively, just make ` MessageSafe` class which inherits `Message` and deletes `get`/`put`.

Also, I don't immediately understand why these intrusive ptr functions are necessary. Shouldn't the `RefCountedObject` versions be used by the compiler?

Reply:
> just make ` MessageSafe` class which inherits `Message` and deletes `get`/`put`.

i added a `SafeMessage`. as we need to have a single-root inheritance architecture for Message classes, and i don't really want to templaterize this helper class.

> Also, I don't immediately understand why these intrusive ptr functions are necessary. 

@batrick yeah, they are not necessary. i thought they were. but i now realize that a pointer to `MMgrDigest` **is** a pointer to `RefCountedObject` in this context.